### ISSUE TITLE
OOM occurred while running replica GC old version

### ISSUE THREAD

**Issue by heidawei** [cockroachdb/cockroach#42531]
**Describe the problem**

cockroach data node crash due to OOM when a replica running GC.

**To Reproduce**

What did you do? Describe in your own words.
I load data to a table(load like batch insert a large number of rows ) which has auto-increment, after 25hours later, the replica running GC old version for the table, but there are too many version, and when running GC we copy all version into memory. I known auto-increment is not recommended in cockroach, but I think it is a bug.

And I found another problem, when a key has to many version to GC(more than 256KB), it will Increase normal read and write request latency in  that key since the latch.

If possible, provide steps to reproduce the behavior:

1. Set up CockroachDB cluster ...
2. Send SQL ... / CLI command ...
3. Look at UI / log file / client app ...

**Expected behavior**

**Additional data / screenshots**
If the problem is SQL-related, include a copy of the SQL query and the schema
of the supporting tables.

If a node in your cluster encountered a fatal error, supply the contents of the
log directories (at minimum of the affected node(s), but preferably all nodes).

Note that log files can contain confidential information. Please continue
creating this issue, but contact support@cockroachlabs.com to submit the log
files in private.

**Environment:**
 - CockroachDB version [e.g. 2.0.x]
 - Server OS: [e.g. Linux/Distrib]
 - Client app [e.g. `cockroach sql`, JDBC, ...]

**Additional context**
What was the impact?

**Comment by heidawei** [comment]
I try to support a PR to fix this bug, which I had do and test in my environment.
1. fast GC
   when a key has too many old version to GC(> 512KB now) I will use range delete for GC.
2. GC limit
   when the last version + 500ms > time.Now(), we will limit the key versions count to reduce latency.

**Comment by ajwerner** [comment]
> And I found another problem, when a key has to many version to GC(more than 256KB), it will Increase normal read and write request latency in that key since the latch.

Yes, this is known. See https://github.com/cockroachdb/cockroach/issues/17229.

As for the OOM, I think I understand the problem but I'd like to understand how it occurred as I suspect there are easier solutions than your span-oriented GC request. That being said I admire your tenacity in attempting to fix the problem. As I read the code, it seems that we load every version of each individual key into RAM. This seems problematic in the case where there is a very large amount of data underneath a single key. That situation also happens to be exactly the case where a range becomes unsplittable. All that being said, you should receive write backpressure when a range exceeds `kv.range.backpressure_range_size_multiplier`*`range_max_bytes`. 

Did you mess with the range size settings or the backpressure?
Is it the case that you had gigabytes worth of versions of a single key?
If so, how did you get in to such a situation?
What version of cockroach are you running?
Are there details I'm misunderstanding?

Generally if there are just kilobytes of a single key in terms of versions I don't see how you would run of out RAM during GC. Is your machine extremely memory constrained?

I am happy you filed this issue. I'm amenable to (though not enthusiastic about) the idea of using ClearRange to clear many versions of a key. I find #42640 in its current state to be more invasive than it needs to be. As a first step I'd look to paginate the individual versions of a key during the GC process rather than jumping straight to ClearRange. I'd also like to understand how the backpressure mechanisms we have failed to prevent the OOM you encountered before moving forward with any sort of code change.

**Comment by heidawei** [comment]
We use docker to limit the CPU and memory. In general, we use 4 CPU and 8GB of memory by default. In our usage scenario, if I keep writing a table with self-incrementing columns, or if I keep updating a row, then A large number of historical versions will be generated in the table that maintains the self-added column ID, but the fragment of this table cannot be split because there is only one valid key, and the original mechanism will actually generate a write error due to backpressure, but this kind of The scene is very common on our side. Based on business considerations, we must properly correct the logic of the split-related source code. When we identify whether we need to split, we only find that the fragment has only one valid key (range size > 64MB).
), then ignore it, so that it passes through backpressure.

**Comment by heidawei** [comment]
We use docker to limit the CPU and memory. In general, we use 4 CPU and 8GB of memory by default. In our usage scenario, if I keep writing a table with self-incrementing columns, or if I keep updating a row, then A large number of historical versions will be generated in the table that maintains the self-added column ID, but the fragment of this table cannot be split because there is only one valid key, and the original mechanism will actually generate a write error due to backpressure, but this kind of The scene is very common on our side. Based on business considerations, we must properly correct the logic of the split-related source code. When we identify whether we need to split, we only find that the fragment has only one valid key (range size > 64MB).
), then ignore it, so that it passes through backpressure.

**Comment by heidawei** [comment]
The version we are using is V19.1. I used to use another solution to solve the problem of OOM. I used the bucket. With the iteration, I will only keep a limited bucket, and each bucket is small. 1MB, up to 10 buckets, so only the last 10MB of data will be processed, but the test shows that the GC is too slow, because it requires multiple GC schedules, iterates through each time, takes a lot of time, and is easy Cause timeout, in the event of a timeout will encounter another tragedy, that is, there will never be a chance for the GC history version, because the future data will be more and more, vicious circle.
We also tried the way of reverse traversal, but considering that the iterator is not deeply understood, because it has not been used in practice. Later we think that you can use range delete to quickly clean up this extreme scenario (in our business) A large number of outdated versions under the scene are very common.
Here we also propose an improvement, cockroach is simply that all versions are loaded into the memory, but in fact we can filter in advance when loading, which will not be satisfied between the versions that are not satisfied, so you can also reduce the memory The amount of data in.

**Comment by heidawei** [comment]
We are also worried that the introduction of range delete in the GC has some unpredictable risks, so we submitted ISSUE and PR, and we hope to get community discussion and guidance.
In our current tests, our PR solves the impact on normal reading and writing in the GC process, and also solves the long time-consuming problem of eliminating large-scale expired data.
My timeout processing in PR is "hijacking", ignoring timeout, and our internal discussion also feels so strange, but for the time being, there is no better way to deal with this problem, so I hope to discuss this in PR. problem.
In short, we hope to have a more elegant way to solve the problems we face.

**Comment by ajwerner** [comment]
> A large number of historical versions will be generated in the table that maintains the self-added column ID

Oh! I finally now understand that this is caused by our SQL sequence type!
What sort of write rate are you talking about in terms of sequence increments per second?

The code today simply increments the value on a single key. I suspect there are better implementations. There are at least implementations which make different tradeoffs. One such might me to keep sequences as a single row table where the primary key gets incremented every time, deleting the old row. This will lead to much saner GC behavior. The downside is that it can't be done in a single request and becomes quite exposed to retries. Perhaps we could add lower level support to make incrementing cheaper. A hybrid option might to only store so many underneath a given key but need to do an initial read to figure out the key and optimistically assume it doesn't change. Maybe every 16k or something.

Another trick would be to dramatically turn down the GC TTL for the sequence table. What do you have it set to?

> Cause timeout, in the event of a timeout will encounter another tragedy, that is, there will never be a chance for the GC history version

:) I was just looking into timeouts from the queues today... https://github.com/cockroachdb/cockroach/pull/42686

I can easily extend that change to include a knob to control the timeout of the GC queue processing.

>  I used the bucket. With the iteration, I will only keep a limited bucket, and each bucket is small. 1MB, up to 10 buckets, so only the last 10MB of data will be processed, but the test shows that the GC is too slow, 

I don't think I understand this part. Say you did one 10MB batch of versions. You're saying that if you did that continuously until they were all gone, you wouldn't be able to keep up? Or maybe you would be if you hadn't accumulated so much garbage in the sequence table?

Short term I think we should do the following:

* Make a knob to control GC queue process timeout.
* Make a patch to paginate the versions of key if the data size is too large.

At a certain size a clear range might be more efficient but I'd generally be pretty stunned if the GC logic couldn't keep up with increments if the GC TTL was set low enough. It's possible that we should default the GC TTL for sequences to a low value though I can see reasons not to. 

As I come to better understand you use case and how the above steps might be insufficient then we can talk about more ideas.

**Comment by heidawei** [comment]
The use of buckets means this. For example, if there is a 1GB outdated version that requires GC, I only keep the last 10MB of data processing in the memory. The use of the bucket allows me to discard the data generated by the previous iteration. It is not the key to use the bucket. The key is that in order to avoid OOM, for each key, a GC process will only delete up to 10MB of data, so that a round of GC can not complete all the overdue data cleanup, must wait for the next round, so that the overall need for multiple rounds of scheduling Finished, every RunGC needs to traverse the entire range from scratch, causing duplicates.

**Comment by heidawei** [comment]
we can set the sequence table GC TTL, but this is time consuming and laborious. Every time you create a table, you need extra work. We hope to avoid this kind of manual intervention. It is often easy to make mistakes.

**Comment by heidawei** [comment]
If we can solve the problem of the sequence table GC, many times we don't have to keep so many versions. In this case, the probability of the problem we are discussing is very low, only extreme cases.
I think of a way, can we default the sequence table GC TTL is very small? I understand it is a special table, the impact on the backup is very small, other things should be nothing to pay attention to.

**Comment by heidawei** [comment]
Can we keep a mechanism like Fast GC, when users encounter similar extreme situations, users can quickly deal with it through this special channel, because this deteriorating scene is difficult to improve, we have tested and ignored GC timeout It took a day to use the original GC logic, but the effect is not ideal, the range size is still very large, if the business is still written, then this situation will not get better, because each time the GC costs a lot Iterate over all the keys, but only delete the last 10MB (for a key).

**Comment by ajwerner** [comment]
> The use of buckets means this. For example, if there is a 1GB outdated version that requires GC, I only keep the last 10MB of data processing in the memory. The use of the bucket allows me to discard the data generated by the previous iteration. It is not the key to use the bucket. The key is that in order to avoid OOM, for each key, a GC process will only delete up to 10MB of data, so that a round of GC can not complete all the overdue data cleanup, must wait for the next round, so that the overall need for multiple rounds of scheduling Finished, every RunGC needs to traverse the entire range from scratch, causing duplicates.

I think that you misunderstand what I mean by pagination. The idea is not to allow the RunGC method to return until all of the garbage has been removed. The RunGC method already subdivides it’s processing into several requests if the data is due to multiple keys. I’m not suggesting only clearing 10MB per gc queue process cycle. I’m suggesting clearing as much as possible during a gc queue cycle but with 10 MB worth of data scanned at a time. The iterator state can be preserved avoiding a rescan. 

I agree that if you could only clear 10MB per gc process it would be a problem. That problem would be made much worse if the gc ttl is 25 hours. 


> If we can solve the problem of the sequence table GC, many times we don't have to keep so many versions. In this case, the probability of the problem we are discussing is very low, only extreme cases.
> I think of a way, can we default the sequence table GC TTL is very small? I understand it is a special table, the impact on the backup is very small, other things should be nothing to pay attention to.

It sounds like providing a mechanism to have a short default gc ttl for sequences is a good idea and seems like it would have prevented your problem in the first place.

**Comment by ajwerner** [comment]
> Can we keep a mechanism like Fast GC, when users encounter similar extreme situations, users can quickly deal with it through this special channel, because this deteriorating scene is difficult to improve, we have tested and ignored GC timeout It took a day to use the original GC logic, but the effect is not ideal, the range size is still very large, if the business is still written, then this situation will not get better, because each time the GC costs a lot Iterate over all the keys, but only delete the last 10MB (for a key).

That “special channel” is something I actively would like to avoid. I get your point that ClearRange can be a more efficient way to delete many versions of the same key. Let’s see about making the existing per-key delete logic more robust to multiple versions before we go and do anything extreme.

**Comment by heidawei** [comment]
I will also conduct more tests and attempts to fix this problem. The ideas and principles you provided give a new perspective to solve the problem, thank you very much
